src/hooks.rs

```diff
@@ -132,8 +134,71 @@ pub fn register_format_detection_hook(
     signature: &'static [u8],
     mask: Option<&'static [u8]>,
 ) {
+    let extension = extension.to_ascii_lowercase();
     GUESS_FORMAT_HOOKS
         .write()
         .unwrap()
         .push((signature, mask.unwrap_or(&[]), extension));
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{ColorType, ImageReader};
+    use std::io::Cursor;
+
+    const MOCK_HOOK_EXTENSION: &str = "MOCKHOOK";
+
+    struct MockDecoder {}
+    impl ImageDecoder for MockDecoder {
+        fn dimensions(&self) -> (u32, u32) {
+            (1, 1)
+        }
+        fn color_type(&self) -> ColorType {
+            ColorType::Rgb8
+        }
+        fn read_image(self, _buf: &mut [u8]) -> ImageResult<()> {
+            panic!("MOCK decoder called")
+        }
```

Comment:
I would avoid using panics as a test mechanism wherever possible. We want to test here that a decoder associated with a specified hook was executed. I think a nicer way to check would be using an atomic bool or `std::sync::Once`:

```
use std::sync::{Arc, Once};

struct MockDecoder {
    done: Arc<Once>,
}
```

```suggestion
        fn read_image(self, _buf: &mut [u8]) -> ImageResult<()> {
            self.done.call_once(|| {});
        }
```

```
let hook_verifier = Arc::new(Once::new());

register_decoding_hook(
    MOCK_HOOK_EXTENSION.into(),
    Box::new(move |_| {
        let decoder = MockDecoder { done: hook_verifier.clone() };
        Ok(Box::new(decoder))
    });
);
```

```
ImageReader::open("tests/images/hook/extension.MoCkHoOk")
            .unwrap()
            .decode()
            .unwrap();

assert!(hook_verifier.is_completed());
```

Reply:
ahh thanks! I was trying to figure out how to do that nicely.  I'll apply these changes.

Reply:
That's also a way to communicate the use explicitly, I like it.